CIP-0130? | Transaction Pieces (#873)
cardano-foundation/CIPs: CIP-????/README.md

---

> CIP: ????

Reviewer: ```suggestion
CIP: 130
```
Please also retitle containing directory to `CIP-0130` and update the rendered proposal link in your OP 🎉

Title: Transaction Pieces
Status: Proposed
Category: Ledger
Authors:
    - fallen-icarus <modern.daidalos@gmail.com>
Implementors: []
Discussions:
    - https://github.com/cardano-foundation/CIPs/pull/862#discussion_r1700138228
    - https://github.com/cardano-foundation/CIPs/pull/862#issuecomment-2266689423

> Solution-To: 

Reviewer 2: ```suggestion
```
adding to the header will break downstream tools

single piece can encompass multiple inputs, outputs, mints, etc) and *they can be unbalanced*. Being
able to create unbalanced intents is critical to a healthy intent ecosystem because most intents are
fundamentally only pieces of a transaction. Look back at Alice's and Bob's pieces above. The atomic
swap intents don't say anything about the counter-party to the swap, and they shouldn't need to.
Requiring users to specify who the counter-party should be fundamentally limits who can actually
fulfill the intent. Without being able to create unbalanced transaction pieces, users would be
forced to specify inputs and outputs for the potential counter-party just to finish balancing the
transaction. 

Another benefit of not having to specify counter-party information is that these transaction pieces
can be trivially batched together without requiring coordination. As a result, a single transaction
can trivially satisfy the intents of many complete strangers. In short, being able to properly model
unbalanced intents makes transaction pieces a very natural and efficient intent settlement system.

> > [!IMPORTANT]

Reviewer: ```suggestion
> ⚠️ **Important**<br>
```
This looks great on GitHub there are sites derived from the CIP source material where this construction might not look very good, or at least will appear unstyled.  Currently no merged CIPs use this construction & you might be the first. 

Maybe you can tell from https://github.com/cardano-foundation/cf-cip-frontend how it would look (I searched for "admonition" on the source code & didn't find anything).  All things considered it might be suitable to leave these the way they are, but in general I think someday we will want CIPs with portable primitives (i.e. classic Markdown + the math extensions).

Reviewer: @fallen-icarus I'm going to withdraw this suggestion since for some reason the CIP author community has started to converge upon GitHub admonitions recently; e.g.: https://github.com/cardano-scaling/CIPs/blob/decentralized-message-queue/CIP-0XXX/README.md#mithril-extra-network-usage

If & when it appears awkwardly on derived sites then I'll argue it's those sites' obligation to add some support for this GitHub syntax & display it intelligently.  Please keep an eye on how this looks on the Dev Portal and `cips.cardano.org` after it is merged & raise a red flag if the admonitions make a poor showing there. :)

#### Transaction

Below is the current cddl representation of a transaction:

```cddl
transaction =
  [ transaction_body
  , transaction_witness_set
  , bool
  , auxiliary_data / null
  ]
```

It contains a single transaction body and a single witness set. This will be changed to contain a
list of transaction bodies, each paired with their own witness set.

The transaction witness set is currently:

```cddl
transaction_witness_set =
  { ? 0: [* vkeywitness ]
  , ? 1: [* native_script ]
  , ? 2: [* bootstrap_witness ]
  , ? 3: [* plutus_v1_script ]
  , ? 4: [* plutus_data ]
  , ? 5: [* redeemer ]
  , ? 6: [* plutus_v2_script ] ; New
  }
```

Smart contracts will still be executed against the full transaction (this will be discussed in a
later section), but all of the required information for an execution (eg, script, datum, and
redeemer) can be found within the corresponding transaction piece. This enables the witness set to
remain unchanged.

The new transaction would then be:

```cddl
transaction_piece =
  { 0: transaction_body
  , 1: transaction_witness_set ; The witness set for this transaction body.
  }

transaction =
  [ transaction_pieces    : [* transaction_piece]
  , bool

>   , auxiliary_data / null

Author: I thought I mentioned that, but I guess I accidentally deleted that part while editing. Each piece *can* specify it, but if multiple do, they must all agree. I wasn't anticipating users to use pieces for storing metadata on-chain. I thought they would build a personal transaction for that.

>   , auxiliary_data / null

Reviewer 3: So it is the "must agree" thing sprinkled throughout the CIP that I really have a problem with. Because of those agreements would have to be enforced by the ledger rules. Which increases complexity of already very complex rules.

  ]
```

The order of the transaction pieces list is significant since it represents the order in which the
pieces must be combined, and ultimately determines the order of the inputs and outputs. How to

> combine the pieces will be discussed later in the spec. 

Reviewer 3: This is very true. However, I can't find any security mechanism that would preserve the order of transaction pieces. There is no guarantee that man in the middle or a nefarious pool will not reorder the transaction pieces in their favor

   ;    length of transaction_bodies
```

Just like with the transaction cddl, the block's transaction bodies and the transaction witness sets
will be merged, and the constraints must be updated accordingly:

```cddl
; This is re-posted here for reference.
transaction_piece =
  { 0: transaction_body
  , 1: transaction_witness_set ; The witness set for this transaction body.
  }

; A specific transaction in the block is a specific batch of transaction pieces. 
block_transaction = [* transaction_piece]

block =
  [ header
  , block_transactions   : [* block_transaction] ; This is new
  , auxiliary_data_set   : {* transaction_index => auxiliary_data }
  , invalid_transactions : [* transaction_index ]
  ]; Valid blocks must also satisfy the following constraint:
   ; 1) every transaction_index must be strictly smaller than the
   ;    length of block_transactions
```

### Transaction Size Limit

A transaction broken into pieces should still have approximately the same total size as the whole
transaction. There are a few things to note:

- If two transaction pieces require the same witness (ie, both pieces require a signature from

> Alice), this witness would now effectively appear twice.

Author: Yes, that is what I meant by "effectively", but I could have been more explicit.

- Certain items can appear in multiple pieces (eg, the same input can be spent/referenced, tokens can
be minted/burned that would otherwise cancel out, etc).

Despite the above, it is probably fine to leave the maximum transaction size as is, at least to
start. The transaction fee would naturally incentivize users/aggregators to minimize the amount of
duplication in a transaction.

### Block Size Limit

Due to the possible duplicate items in each transaction and the fact that a block is made up of
multiple transactions, the issues in the Transaction Size Limit section are slightly exacerbated.
However, it should still be fine to keep the current block size limit, at least initially. Just like
with transactions, the transaction fee would naturally incentivize users to minimize duplications.

### Smart Contract Execution

While the cryptographic keys only need to sign the relevant transaction pieces, the smart contracts

> would be executed against the full transaction. This requirement enables the current smart contract

Author: > Without a good reason for this, it's probably better to properly reflect this structure in `TxInfo`.

Personally, whichever option minimizes added complexity is my preferred choice. If we preserve the transaction as you are suggesting, the smart contract interface would need a pretty large change and DApps would need to be re-architected. Perhaps keeping the structure in `TxInfo` would be better in the long run, but I'm not sure. I do not know what changes would be required on the ledger/plutus side to actually preserve the structure in `TxInfo`.

@vlasin If we did preserve the structure, I believe this would address your [suggestion](https://github.com/cardano-foundation/CIPs/pull/873#issuecomment-2270494230).

interface to remain mostly untouched. The flow is essentially:

> 1. Deterministically derive the whole transaction from the list of pieces.

Author: > - two transactions that withdraw from the same reward account. It is irrelevant that one of them will be zero, it will change semantics. Moreover if a stake credential is a scritp, then how can you ignore running one of them.

I would argue this transaction should be rejected. I think you are over-complicating things. IMO things should either exact match or be rejected. The only exceptions are when there is a very obvious way to merge them, like with mints and validity intervals. 

> - A DRep casting a yes vote on a proposal in one transaction and then a No on the same proposal in a separate transaction. You might think that we could then ignore the first vote, since the second one replaces the first. But, what if the DRep is backed by a script and it would actually fail if we didn't ignore it.

This transaction would be rejected as well because the proposal vote in one piece does not exact match the proposal vote in the other. The pieces are not applied to the ledger sequentially. They are folded into a single transaction and applied as a unit.

> - Two transactions minting the same asset. How do you run the script twice? You'll be forced to put squash those two mints together into one, which could affect the execution of the script.

The mints would be merged upon folding the pieces and the script would only be executed once. As I said before, smart contracts are a weak-point with this CIP. AFAIU, the new Validation Zone revision addresses this limitation by actually having smart contracts validate against the piece they are used in.

> 1. Deterministically derive the whole transaction from the list of pieces.

Reviewer 3: I am not over-complicating! Those are all of the cases that would have to be addressed in the ledger rules. If we are to reject them, we need to implement specific predicate failures for each one of them that do the rejection. I can see this CIP adding more predicate failures than the total number of predicate failures that we already have today.

> Validation Zone revision addresses this limitation 

I am really confused here. [Validation Zones](https://github.com/cardano-foundation/CIPs/pull/862) is a totally different CIP. Are you trying to combine the two together somehow?

The only way I currently see combining multiple transactions into one for validation to work is if we restrict all those pieces to be **completely** disjoined, i.e. no common credentials.

> 1. Deterministically derive the whole transaction from the list of pieces.

Author: > I am not over-complicating!

Sorry, my confusion came from not realizing the redeemers included the execution budgets and therefore you can't do an exact match on the script_data_hash across pieces. I was under-complicating it :sweat_smile:

> I am really confused here. Validation Zones is a totally different CIP. Are you trying to combine the two together somehow?

The new Validation Zone revision (the [README-implicit.md](https://github.com/cardano-foundation/CIPs/blob/1fd1d6dc56fadca8b07009e795b3ec4eedc298cd/CIP-0118/README-implicit.md)) is trying to incorporate the unbalanced transaction idea from this CIP. It uses them in an entirely different way than how this CIP used them.

2. Generate the script context for the whole transaction.
3. Evaluate the smart contract using this script context.

When combining the transaction pieces, the required executions (ie, script + datum + redeemer) can
be built up as well. Duplicate executions should be removed. For example, if two transaction pieces
look to spend the same smart contract input using the same redeemer, there is no reason to evaluate

> this scenario twice since they are guaranteed to have the same result.

Author: > How can you spend the same input twice?

You wouldn't. You would remove duplicates when merging the pieces. If you have a list `[1,2,3,4]` and another `[2,5,6]` merging them (and removing duplicates) would give `[1,2,3,4,5,6]`. So input `2` is *listed* twice, but only *spent* once.

Smart contracts are a definite weak-point of this CIP, but the new Validation Zones revision seems to address this weak-point while still incorporating the main ideas from this CIP. I personally prefer the new Validation Zones CIP over my own for this reason.

> this scenario twice since they are guaranteed to have the same result.

Reviewer 3: > You would remove duplicates when merging the pieces

So, as I said you can't remove duplicates because of smart contracts.

And with validation zones you can't have duplicates.

> 

Reviewer 4: This looks like a very promising design indeed! I am curious to understand how minting NFTs that require the spending of a specific UTxO entry would work, if two pieces are trying to mint the same NFT? Would you be able to locally test if your NFT contract validates?

Author: > I am curious to understand how minting NFTs that require the spending of a specific UTxO entry would work, if two pieces are trying to mint the same NFT? Would you be able to locally test if your NFT contract validates?

Hmm, I hadn't thought about testing just a piece, but it should be possible since a transaction piece is just a transaction body. This would just require a smart contract to be able to run against an unbalanced transaction body. Whether the smart contract would still succeed when aggregated with other pieces depends on the smart contract logic. So Alice can verify her piece will successfully validate *before* she signs, and the aggregator can verify it will still validate *after* adding it to the batch.

Reviewer 4: In general, it opens up the possibility of an attack if the outcome of contract validation can be changed after you submit your transaction or transaction piece. I see two ways to deal with this :

1. restrict your design to only allow combining pieces that contain no Plutus smart contracts whatsoever

2. simplify Validation Zones to be more in line with what you are suggesting/be more implicit with value flow :
- revert changes to the UTxO rule, UTxO State, and Transaction data structures (leave them the same as in the current non-zone ledger design)
- remove new Plutus version requirement
- move the `produced = consumed` computation to the ZONE rule, and aggregate across all transactions in the zone, that is `Sum_{tx \in zone} consumed tx = Sum_{tx \in zone} produced tx`
- replace the cycle detection in the ZONE rule by in independency check, which checks that "for all inputs `(txid, ix)` in all transactions in a zone, `txid` is not the ID of another transaction in the zone
- only require the per-transaction component of the transaction fee to be paid once per zone

The validation zones simplification will all remove the need for the weird `requireTxs` song and dance, and an aggregator would be able to spend their liquidity pool once per zone to provide any missing value and claim any extra value within a single transaction. Scripts would operate without any changes. Flash loan attacks would be prevented by the independency check.

Reviewer 4: 1. The difference between an account-based dApp doing things and the Boolean outcome of an EUTxO contract is that the dApp can (and has to) be written in a way that accommodates any possible account state to which it can be applied, eg price changes etc. An EUTxO contract fee cannot even be calculated without the complete fixed transaction data.  You can make any contract fail by bulking up the tx size. If you make a transaction that is valid, someone can always make your scripts fail by submitting it batched with whatever nonsense. To avoid this, you have to be able to prevent your transaction from being modified (this is currently possible only by signing it with a key that is required to sign it). Note that even if you yourself didn’t include collateral, and are relying on someone else to include it, the transaction batch they construct can also be messed up. This brings me to the Validation Zones modification - 

2. The change I propose actually allows you to model unbalanced intents by allowing whole transactions that are unbalanced, while the zones containing transactions must be balanced. It is somewhere in between the original design and what you propose. Instead of requests and fulfills, it simply allows assets to be missing from or be extra in a transaction. The preservation of value at the zone level ensures that regardless of who provides the missing assets and claims the extra ones , things balance out. Consider zone [Tx1; Tx2]

Tx1 - 
(Resolves) inputs: txIn -> (10 Ada, …)
Outputs:(5PolinaCoin, …)

Tx2 - 
(Resolved) inputs :txIn’ -> (5PolinaCoin, …)
Outputs :(10Ada, …)

The important thing is that txIn’ is not hash (Tx1), because of flash loans. No need to request 5 PolinaCoin or make an unlocked offer of 10Ada.

Reviewer 5: The weakening to transaction balancing seems like a good middle-of-the-road approach. It means you can optionally have interdependencies but aren't forced to exactly specify everything. I suspect that in practice all it really adds is a minor optimization on transaction size, but that's nice as well.

Author: I have a few questions:

- Transaction Fees - is it possible for unbalanced transactions within a zone to have a fee of zero? Particularly for babel fees, only one transaction in the zone needs to cover the fee for the entire zone. Multiple transactions in the zone could specify the fee (to pay their fair share) in which case the total fee paid would be the sum of all of the fees paid in the zone.
- Flash loans - I always thought flash loans implied no time passing; you borrow a loan and immediately pay it back "before the world notices". But that doesn't seem to be possible here since time is implied to pass: tx1 is processed before tx2. How is this any different than what currently happens where one transaction is processed after another in the same block? The only issue I can see is that the transaction fee would be lower now (assuming it is 0.15 ADA per zone now instead of 0.15 ADA per transaction). I don't see why `txIn'` can't be `= (hash (Tx1), _)`. If it could, it seems trivial to do an atomic batch of transactions since they are intrinsically linked.

Reviewer 4: - i think it will not be possible to specify a fee of 0 (this is because the amount of fee specified in each transaction must go to the fee pot, and that's how the system knows how much to add), but it does not mean you have to provide the fee. Your transaction can simply be short that amount, and someone else's transaction needs to have that amount extra to make a valid zone

- the flash loan problem here really has to do with smart contracts : you should not be able to use a thread token or a role token in your transaction unless you spend an actual output containing one. If we allow transactions in a zone to be chained in this way, you dont actually need to spend a real output with a role token to use one, flash-loaning yourself one :

Tx1 -
(Resolved) inputs : txIn -> (out1)
Outputs: (1RoleToken, …) ,  (out2)

where out1 and out2 are some outputs containing the state of some contract which can only be updated if the transaction contains a RoleToken that allows such an update to be made

Tx2 -
(Resolved) inputs : (hash (Tx1), 1) -> (1RoleToken, …)

Author: > you should not be able to use a thread token or a role token in your transaction unless you spend an actual output containing one.

Perhaps I misunderstood what you meant by `(hash (Tx1), _)`. I may also be misunderstanding what you mean by
> (Resolved) inputs: txIn -> (out1)`. 

The way I am thinking about it, the role token requirement is guaranteed to be true; there is no way to NOT spend the role token. No "resolving" is required.

A zone is just a list of transaction: `[tx1, tx2, tx3]`. But we can force a dimension of time on the list:
```
---- Time ---->
[tx1, tx2, tx3]
```

If we do this, then the above is just a mini-blockchain, and we can treat it as such. The zone can build up its own mini UTxO set, and transactions in the zone have access to the current state of `real UTxO set + zone UTxO set` at the time they are being processed. The zone can keep track of both the UTxOs spent from the real UTxO set and the UTxOs created/spent from the zone UTxO set. When all transactions in the zone have been fully validated, the net affect of the real UTxO set + the zone UTxO set is what actually gets applied to the blockchain.

In the above zone, `tx1` gets processed first. Imagine if `tx1` spends a real output `r1` and produces an output `z1`. The zone would log that `r1` has been spent from the real UTxO set and `z1` has been added to the zone UTxO set. `tx2` comes after `tx1` and can spend any remaining UTxO in the real UTxO set and the zone UTxO set. So `tx2` cannot spend `r1` because it was already spent by `tx1`, but it can spend both `r2` (another input from the real set) and `z1` to produce `z2`. `tx2` would call `z1` explicitly by output reference: `(hash(tx1), z1_index)`. `z1_index` is its index in `tx1`'s output list.

Now the new state is the real UTxO set minus `r1` and `r2`, plus the zone UTxO set of `z2`. `tx3` comes after `tx2` so it must use this new state. If it wants to spend `z2`, it must call it explicitly by output reference: `(hash(tx2), z2_index)`. `z2_index` is its index in `tx2`'s output list. Let's say `tx3` spends `z2` and produces `z3`. `tx3` cannot spend `z1` because it was already spent by `tx2`.

The net effect of this zone is `r1` and `r2` are both spent to produce `z3`. This is the change that would actually be recorded on the blockchain. 

With this method, if the role token was in `r1`, `tx1` moved it to `z1`. If `tx2` wants to execute the smart contract that requires the role token, it must now spend `z1` because that is where the token is at the time `tx2` is being processed. If `tx3` wants to use the role token as well, it must spend `z2` since that is where `tx2` moved the token to. Because time is enforced on the zone, the zone can know where the role token is at all times. When `z3` is actually recorded on-chain, it is the result of `r1` -> `z1` -> `z2` -> `z3`. The eUTxO model guarantees the result is the same as doing the updates separately (ie, in separate zones).

> [!Important]
> With this approach, building a transaction chain **requires** coordination between the transactions in the chain. However, I am making the assertion that: *coordination is always possible whenever transaction chaining is useful*.  If we accept this assertion, there is no need for any "resolving", and flash loans are impossible.

Strangers can't build up chains this way since they don't know each others' transactions (and therefore each others' output references), but why would strangers be chaining their transactions anyway? This is mainly useful for account style DApps that are controlled by batchers.

A batcher can personally build up the chain when trying to fulfill the user orders. They can create `tx1` first and use its hash in `tx2`, and then use `tx2`'s hash in `tx3`. Furthermore, since the batcher knows the order of the outputs in each transaction, they know what the resulting output references will be in advance. The batcher can even turn the user orders into chains (if the user transactions allow for it). If batchers need to build up a chain together, they can coordinate this off-chain.

Because the outputs are being called explicitly by output reference, it is trivial to check if the role token is actually being spent. The smart contracts don't even need to know they are being executed in a zone.

Author: > i think it will not be possible to specify a fee of 0 (this is because the amount of fee specified in each transaction must go to the fee pot, and that's how the system knows how much to add)

I don't see why it is possible for the zone to know if it is balanced, but not know if it has enough ADA provided for the fee. The zone checks should have everything they need to calculate the total required fee for the zone, and it can compare this against the sum of the fee specified in each transaction within the zone. This would enable some transactions to have a fee of zero while still being part of a valid zone.

Reviewer 4: > I don't see why it is possible for the zone to know if it is balanced, but not know if it has enough ADA provided for the fee. 

This is maybe possible? Removing the per-transaction fee check is likely possible, and if we do, entering a paid fee of 0 for a transaction would also be possible. I need to think more about this, though, and would be happy to discuss in CIP meeting.

I think i need a diagram for your previous comment ...

Here is a draft of the "Implicit Value Flow" validation zones version, it is much simpler than the original

https://github.com/cardano-foundation/CIPs/pull/862/commits/1fd1d6dc56fadca8b07009e795b3ec4eedc298cd#diff-560051716ec259fa49b3ef402bb6c1f4b7517d1bf400db3e735996986398bf96

Reviewer 4: @AndrewWestberg @WhatisRT  if you feel like having a look also

waste execution resources.

### Transaction Piece Monoid Instance

The monoid instance for transaction pieces defines the rules for combining the pieces. In order to
be able to deterministically derive the full transaction, this instance must be standardized. The
monoid instance requires answering two questions:

1. What is an empty transaction piece?
2. How are two transaction pieces combined?

The first question is simple to answer: an empty transaction piece is an empty transaction body.
Users fill out as much information as they wish when creating their transaction piece. The rest of
the fields can remain empty.

> [!IMPORTANT]
> It doesn't really make sense to create a transaction piece with an input but no output, or an
> output but no input. Therefore, these fields can remain required fields. The fee is also required,
> but that can be set to zero in the pieces not paying the fee. All other fields in the transaction
> body are already optional.

The second question requires more consideration, but as a general rule, as long as there are no
incompatibilities between transaction pieces, they can be combined into the full transaction. Any

> transaction that contains incompatible pieces should be rejected.

Reviewer 3: Any two transactions that have plutus scripts would be immediately incompatible.

Imagine a script that relies for there to be only a single input being spent or a single output being produced. As soon as you combine two transactions together they no longer validate. However, you can't know that until you try to combine them and run the scripts

There is another fundamental problem here. Imagine those inputs in two separate transactions are both locked by scripts. Those two transactions both provide script integrity hash that have potentially been signed. But whenever you combine those two transactions together you must change their redeemers and pass those redeemer modified in the script context to plutus scripts, which is totally screwing with the data that have been signed.

So, combining multiple pieces into one transaction idea is definitely out of the window

For example, if Alice creates a piece with `input1` followed by `input2` and Bob creates a piece
with `input3` followed by `input4`, then Charlie (the aggregator) only has two options:

```
[input1, input2, input3, input4] OR [input3, input4, input1, input2]
```

Charlie cannot put `input2` before `input1` because this would violate Alice's intent. The order of
everything else is standardized (eg, token mints are sorted by name, withdrawals are sorted by
credential, etc), so no one can control the order of them.

> [!IMPORTANT] 
> This would required [CIP-128](https://github.com/cardano-foundation/CIPs/pull/758) since the order
> of the inputs and reference inputs would need to be preserved. There may be intents where the
> order of these fields matter which is why the intent creator should get to specify their ordering.
> The collateral inputs could be an exception, but if we are already making the change for inputs
> and reference inputs, why not also make the change for collateral inputs?

#### Duplicate Inputs

It is valid for two separate transaction pieces in a batch to spend the same UTxOs (eg, a liquidity

> source UTxO). This UTxO can still only be spent once and the batch must still be fully balanced

Reviewer 3: This not valid. Their position in the input set can affect the execution outcome of a plutus script that locks them

> transaction piece needs to specify a transaction fee amount.

#### All Other Fields

All other fields are essentially just lists and have a standardized sorting method. Therefore, they
should just be concatenated and sorted appropriately. Duplicates should be removed from the
resulting lists.

If the field can require a smart contract execution, redeemer disagreements across transaction
pieces should not be allowed.

### Phase 1 Validation

Phase 1 validation must be slightly changed to check:
1. All of the transaction pieces can be combined into a whole transaction (ie, there are no
   incompatibilities).
2. Each transaction piece is properly witnessed.
3. The whole transaction is a valid transaction. 

The rest of the current phase 1 checks should still be performed against the whole transaction (eg,
properly balanced, pays enough of a transaction fee, etc).

### Phase 2 Validation

Phase 2 validation remains unchanged since the smart contracts are still executed against the whole
transaction. 

> TODO: Does the collateral claiming mechanism (in the case of a phase 2 failure) need to be
> changed? I do not believe so, but I am not entirely sure.

> ## Rational: how does this CIP achieve its goal?

Reviewer: ```suggestion
## Rationale: how does this CIP achieve its goal?
```

to try to guess what actions the counter-party will use when satisfying their intent. In practice,
this means Alice must use input and output **placeholders** in her intent, and these placeholders
must also specify the value of the UTxO that will eventually go there; these placeholders are
necessary to finish balancing the intent transaction. But this creates problems for potential
counter-parties: 

- Only UTxOs that exactly fit those placeholders can be used
- The counter-party can't add or remove placeholders after the fact

Imagine if Alice wants to swap 100 ADA for 50 DJED; this says nothing about how the counter-party
will satisfy her swap. Perhaps Bob (the counter-party) wants to provide the DJED, but it is
currently broken up across three UTxOs. And perhaps he wants the 100 ADA he receives to be stored in
two separate outputs. If Alice specifies only one input placeholder and one output placeholder in
her intent transaction, Bob might be unable to act as her counter-party. If Bob really wanted to
still trade with Alice, he would have to find a way to merge his inputs and outputs into one input

> and one output. This may not always be possible and Bob may have to give up on being Alice's

Reviewer 5: This is actually quite easy to do with validation zones. Bob just needs to spend his three outputs and Alice's Ada output and negative Djed output, and make two new outputs containing the Ada. The only extra effort is to spend Alice's two outputs.

Author: See my other comment [here](https://github.com/cardano-foundation/CIPs/pull/873#discussion_r1705441782).

counter-party. In other words, because Alice was forced to specify placeholders in her intent, she
missed out on a potential trading partner.

DApps would also be unable to be direct liquidity sources for these intents due to not always being
able to match the value specified in each placeholder. If the input place holder is for 10.3 ADA but
the liquidity source UTxO actually has 20.5 ADA, this liquidity source cannot be used despite it
having enough liquidity. The Validation Zones CIP argues that DApps can just add support for
breaking-off/merging UTxOs so that the liquidity source UTxOs can be forced into the proper forms

> for the required placeholders. However, this would dramatically increase the complexity and attack

Reviewer 5: I don't see where this is discussed in the Validation Zones CIP. But I also don't see a big problem here, just by going on with the same logic from the last paragraph. A matching transaction can always spend the 'problematic' outputs and then construct the remainder of the transaction as it would construct the transaction piece. I don't see where this impacts 'dramatically increases complexity'.

Author: > A matching transaction can always spend the 'problematic' outputs and then construct the remainder of the transaction as it would construct the transaction piece. I don't see where this impacts 'dramatically increases complexity'.

I guess I'm just biased. Given there is a path that doesn't require these extra steps, why wouldn't you take that path? Why would you ever take the path of *more* resistance?

in the Validation Zones' design.

In addition to the counter-party issue, there must be a way for the ledger to actually support and
satisfy these placeholders. To accomplish this, the Validation Zones CIP introduces zones, unlocked
requests, and unlocked outputs. These new features require adding a lot of extra complexity to the
ledger and makes using Validation Zones intents a very complicated experience. For example,
satisfying a single intent actually requires multiple transactions (packaged as a zone).

Again, the counter-party issue and extra complexity in Validation Zones are simply due to not being
able to accurately model unbalanced intents. Since this CIP *can* model unbalanced intents, it can
cover all of the same use cases while only adding a fraction of the complexity.

It is also worth noting that, since the Validation Zones design uses otherwise unnecessary
placeholders and extra transactions just to satisfy these placeholders, the impact Validation Zones
would have on the average block size is likely equivalent to, if not larger than, the impact this
CIP would introduce.

> 

Reviewer 5: Compared to the validation zones approach, this approach appears to be somewhat simpler indeed, but it also has some restrictions. For example, this approach doesn't allow for atomic batch transactions in general. A transaction piece cannot spend an output generated by another transaction piece in the same transaction. I see this as a major drawback, and it should be discussed here.

Author: Atomic batch transactions seems orthogonal to intents (it isn't mentioned in the CPS), but I can still discuss it in the CIP.

@polinavino Perhaps there is a way to use transaction pieces as the foundation for zones?

### Westberg's "Smart Transactions"

[Smart
Transactions](https://github.com/input-output-hk/Developer-Experience-working-group/issues/47) is
very similar to Validation Zones in that it tries to use fully balanced transactions to model all
intents. Because of this, it suffers from the same counter-party issue, over-complexity, and
possible overly-increased average block size as Validation Zones.

### Original Babel Fees Design

The original design effectively required minting policies to be executed in every transaction, but
this is not always possible/feasible.

## Path To Active

### Acceptance Criteria

> - [] It is approved by the community and the ledger/plutus teams.
> 
> ### Implementation Plan
> 
> - [] The ledger spec is updated to use transaction pieces.
> - [] Implementation of the Cardano node with a new era/hardfork.
> - [] Deployed on testnet/mainnet

Reviewer: ```suggestion
- [ ] It is approved by the community and the ledger/plutus teams.

### Implementation Plan

- [ ] The Ledger spec is updated to use transaction pieces.
- [ ] Implementation in the Cardano node with a new Ledger era / hardfork
- [ ] Deployed on testnets / Mainnet
```
Fixing typography & some spelling.  I do think we need more detail in _Acceptance Criteria_ but I think these will emerge from discussion points over a long Ledger + community review.
